src/basedtyping/transformer.py

```diff
@@ -0,0 +1,257 @@
+                self._typing("TypeIs"), cast(ast.expr, self.generic_visit(node.comparators[0]))
+            )
+            return self.generic_visit(result)
+        return self.generic_visit(node)
+
+    @override
+    def visit_IfExp(self, node: ast.IfExp) -> ast.AST:
+        if (
+            isinstance(node.body, ast.Compare)
+            and len(node.body.comparators) == 1
+            and isinstance(node.body.ops[0], ast.Is)
+        ):
+            node.body = self.subscript(
+                self._typing("TypeGuard"),
+                cast(ast.expr, self.generic_visit(node.body.comparators[0])),
+            )
+        return self.generic_visit(node)
+
+    def visit_FunctionType(self, node: ast.FunctionType) -> ast.AST:  # noqa: N802 https://github.com/KotlinIsland/basedmypy/issues/763
+        node = self.generic_visit(node)
+        assert isinstance(node, ast.FunctionType)
+        return ast.Expression(
+            self.subscript(
+                self._typing("Callable"),
+                ast.Tuple([ast.List(node.argtypes, ctx=ast.Load()), node.returns], ctx=ast.Load()),
+            )
+        )
+
+    @override
+    def visit_BinOp(self, node: ast.BinOp) -> ast.AST:
+        node = self.generic_visit(node)
+        if isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitAnd):
+            node = self.subscript(
+                self._basedtyping("Intersection"),
+                ast.Tuple([node.left, node.right], ctx=ast.Load()),
+            )
+        return node
+
+    def _transform(self, value: typing.ForwardRef) -> ast.Expression:
+        tree: ast.AST
+        try:
+            tree = ast.parse(value.__forward_arg__, mode="eval")
+        except SyntaxError:
+            arg = value.__forward_arg__.lstrip()
+            if arg.startswith(("def ", "def(")):
+                arg = arg[3:].lstrip()
+            tree = ast.parse(arg, mode="func_type")
+
+        tree = self.visit(tree)
+        assert isinstance(tree, ast.Expression)
+        return tree
+
+
+def eval_type_based(
+    value: object,
+    globalns: dict[str, object] | None = None,
+    localns: dict[str, object] | None = None,
+    *,
+    string_literals: bool,
+) -> object:
+    """Like `typing._eval_type`, but supports based typing features.
+    Specifically, this transforms `1 | 2` into `typing.Union[Literal[1], Literal[2]]`
+    and `(int) -> str` into `typing.Callable[[int], str]` etc.
+    """
+    if not isinstance(value, typing.ForwardRef):
+        return value
+    transformer = CringeTransformer(globalns, localns, string_literals=string_literals)
+    tree = transformer._transform(value)
+    return transformer.eval_type(tree, original_ref=value)
+
+
+if typing.TYPE_CHECKING:
+
+    def _eval_direct(
+        value: object,  # noqa: ARG001
+        globalns: dict[str, object] | None = None,  # noqa: ARG001
+        localns: dict[str, object] | None = None,  # noqa: ARG001
+    ) -> object:
+        ...
```

Comment:
stuff within `if typing.TYPE_CHECKING` is effectively a temporary `.pyi`, so I guess that the formatting should porbably also reflect this

```suggestion
    ) -> object: ...
```

Reply:
take it up with `ruff`

Reply:
the ruff formatter already does this:
https://play.ruff.rs/de7a2894-5896-4ecf-abb8-0239df638d90

Reply:
it'll probably work once you upgrade ruff 🤷🏻 

```toml
[tool.poetry.group.dev.dependencies]
basedmypy = "^2"
pytest = "^8"
ruff = "~0.2.1"
```

Reply:
i need to update the project to uv etc at a later date